CorsixTH/Lua/world.lua

```diff
@@ -2867,3 +2850,64 @@ function World:setCampaignData(campaign_data)
   end
   self:getLocalPlayerHospital():setCampaignData(campaign_data.hospital)
 end
+
+--! Find a valid disease that a patient can use
+--!return disease (table)
+function World:getNewPatientDisease()
+  -- Construct disease, take a random guess first, as a quick clear-sky attempt.
+  local hospital = self:getLocalPlayerHospital()
+  local disease = self.available_diseases[math.random(1, #self.available_diseases)]
+  if not isDiseaseUsableForNewPatient(self, disease, hospital) then
+    -- Lucky shot failed, do a proper calculation.
+    local usable_diseases = {}
+    for _, d in ipairs(self.available_diseases) do
+      if isDiseaseUsableForNewPatient(self, d, hospital) then
+        usable_diseases[#usable_diseases + 1] = d
+      end
+    end
+
+    if #usable_diseases == 0 then return nil end
+
+    disease = usable_diseases[math.random(1, #usable_diseases)]
+  end
+  return disease
+end
+
+--! Choose hospital for normal patient
+--!param disease (table) - the disease assigned to this patient
+--!return hospital  (Hospital)
+function World:chooseHospitalForPatient(disease)
+  -- round robin if less than gbvAllocDelay
+  if self.map.level_config.gbv.AllocDelay and self.game_date:monthOfGame() <= self.map.level_config.gbv.AllocDelay then
+    self.total_population = self.total_population or 0
+    local hospital_index = self.total_population % 4 + 1
+    self.total_population = self.total_population + 1
+    return self.hospitals[hospital_index]
+  end
+
+  -- sometimes randomly choose a hospital
+  if self.map.level_config.gbv.AllocRand and math.random(1, self.map.level_config.gbv.AllocRand) == 1 then
+    return self.hospitals[math.random(1, 4)]
+  end
+
+  -- Select a hospital based on the reputation and cost of treatment of the patients disease
+  local value_proposition = 0
+  local hosp_patient_value = {}
+  for s, hosp in ipairs(self.hospitals) do
+    local rep_multiplier = self.map.level_config.gbv.AllocIndRep and self.map.level_config.gbv.AllocIndRep or 2
+    local reputation = hosp:getDiseaseReputation(disease.id)
+    value_proposition = hosp:getDiseaseReputationPriceFactor(disease.id)
+    value_proposition = value_proposition * rep_multiplier * reputation
+    if s > 1 then
+      value_proposition = value_proposition + hosp_patient_value[s - 1]
+    end
+    hosp_patient_value[s] = value_proposition
+  end
+
+  local threshold = math.random(0, math.floor(value_proposition - 1))
```

Comment:
Nice, though clever enough it deserves a comment (or the comment above expanded) i.e. to include that the hospital is chosen randomly with a chance proportional to it's relative value proposition.

Reply:
Seconding this... I think

Reply:
Yeah the hasStaffedDesk, they don't spawn, but they do count to the visitor count on the annual report but not on the charts/graphs ui, so maybe I should adjust that also. Though I think that will introduce another statistic. Looks like I could repurpose the World.total_population and Hospital.population variables to account for this.